src/mlpack/methods/ann/layer/isru.hpp

```diff
@@ -0,0 +1,183 @@
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class ISRU
+{
+ public:
+  /**
+   * Create ISRU object using specified hyperparameter alpha.
+   * Default (alpha = 0.1). Alpha should be > 0.
+   * 
+   * @param alpha parameter
+   */
+  ISRU(const double alpha = 0.1);
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   * 
+   * @param input Input data used for evaluating the ISRU function.
+   * @param output Resulting output activation
+   */
+  template<typename InputType, typename OutputType>
+  void Forward(const InputType&& input, OutputType&& output);
+
+  /**
+   * Ordinary feed backward pass of a neural network, calculating the function
+   * f(x) by propagating x backwards through f. Using the results from the feed
+   * forward pass.
+   * 
+   * @param input The propagated input activation f(x).
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient
+   */
+  template<typename DataType>
+  void Backward(const DataType&& input,
+                DataType&& gy,
+                DataType&& g);
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the hyperparameter alpha.
+  double const& Alpha() const { return alpha; }
+  //! Modify the hyperparameter alpha.
+  double& Alpha() { return alpha; }
+
+  /**
+   * Serialize the layer.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  /**
+   * Computes the value of ISRU activation function.
+   *
+   * @param x Input data.
+   * @return f(x) 
+   */
+  double Fn(const double x)
+  {
+    return x / (std::sqrt(1 + alpha*x*x));
+  }
+
+  /**
+   * Computes the value of ISRU activation function using a dense matrix
+   * as input.
+   * 
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename eT>
+  void Fn(const arma::Mat<eT>& x, arma::Mat<eT>& y)
+  {
+    y = x / (arma::sqrt(1 + alpha * arma::pow(x, 2)));
+  }
+
+  /**
+   * Computes the inverse of the ISRU function
+   * 
+   * @param y
+   * @return f^{-1}(y)
+   */
+  double Inv(const double y)
+  {
+    return y / std::sqrt(1 - alpha * y * y);
+  }
+
+  /**
+   * Computes the inverse of the ISRU function
+   * 
+   * @param y Input data.
+   * @param x The resulting inverse of the input data
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Inv(const InputVecType& y,
+                  OutputVecType& x,
+                  const double alpha = 0.1)
+  {
+    x = y / arma::sqrt(1 - alpha * arma::pow(y, 2));
```

Comment:
Could you also the condition if  y ==1 / alpha^{2} so that it becomes DBL_MAX.

Reply:
I believe you mean, if y == 1 / alpha^{1/2}, because that is the case when division by zero occurs.

Reply:
yes, to prevent division by zero.